src/components/annual-report/report-data/section2/section2dCalculations.js

```diff
@@ -0,0 +1,158 @@
+import { convertAnalyticsResponseToObject, getVal } from '../utils/utils.js'
+import { getForecastValue, getMean, getRoundedValue } from './mathService.js'
+import {
+    OVERALL_ORG_UNIT_SECTION_2D,
+    LEVEL_OR_GROUP_SECTION_2D,
+} from './useFetchSectionTwoData.js'
+
+const get2dScore = ({
+    ou,
+    dx,
+    trend,
+    response,
+    currentPeriodID,
+    comparisonPeriods,
+}) => {
+    const currentPeriodValue = Number(
+        getVal({ response, dx, pe: currentPeriodID, ou })
+    )
+
+    // need to keep relative distance for forecast (hence get index, then filter out invalid values)
+    const comparisonPeriodPoints = comparisonPeriods
+        .map((pe) => Number(getVal({ response, dx, pe, ou })))
+        .map((val, index) => [index, val])
+        .filter((point) => !isNaN(point[1]))
+
+    if (trend === 'constant') {
+        const comparisonPeriodValues = comparisonPeriodPoints.map(
+            (point) => point[1]
+        )
+        return (currentPeriodValue / getMean(comparisonPeriodValues)) * 100
+    }
+    // else forecast
+    return getForecastValue({
+        pointsArray: comparisonPeriodPoints,
+        forecastX: comparisonPeriods.length,
+    })
```

Comment:
I haven't gotten to test this because I haven’t found any data items with expected increasing or decreasing trends, but I think the `forecastX` isn't quite right here — I think if `comparisonPeriods` is in order of increasingly ‘old’ periods, then `forecastX = comparisonPeriods.length` will represent be the oldest year, before the comparison periods, rather than the current period

As it’s implemented now, I think `forecastX` should be `-1` (if the year before the current period is x=0, 2 years before is x=1, and 3 years before is x=2) -- but there’s also the option to increase the indices/x-values of the reference years by 1 in [line 23 above](https://github.com/hisprwanda/who-data-quality-annual-report/blob/7962c7aa3b8d790b25f2b28bfb4030a6034c66cb/src/components/annual-report/report-data/section2/section2dCalculations.js#L23), which would make `forecastX` conveniently 0 🙂

If you do that, it simplifies `getForecastedValue()` a bit since the forecasted value would be just the y-intercept of the regression line, i.e. `forecastedValue = forecast.equation[1]` 😀

Reply:
Thanks! 🤦‍♂️

I originally had the periods hardcoded in ascending order [2019,2021,2022], and forgot to update this.

I like the proposal for structure, but I'm going to keep the index values of the initial x period 0 (consistent with the old app) as it might make it easier to interpret for others if the regression for the forecast period is the last number (as opposed to the mathematically convenient equivalent form of setting the forecast-x to zero and flipping the slope)

I also noticed that:

- I was returning the forecast value (rather than the ratio of the current/forecast)
- the forecast value in the old app was floored at zero (in case the forecast was negative). While I'm not sure about the mathematical soundness of this, I assume that it was a deliberate decision (potential follow up for Olav).

I've updated the configurations for one of the underlying numerators (ANC 1 Visits) to use 'increasing' trend (in old and new datastores). Ideally, we'd have a test set that covers the full range of covered cases, but we need as a minimum to have this one covered.